### PR TITLE
Больше монет богу монет!

### PR DESCRIPTION
**Pull request by RayProud**
---
:mortar_board: [Больше монет богу монет!](https://up.htmlacademy.ru/nodejs/2/user/22908/tasks/15)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,8 +27,10 @@
     "supertest": "3.3.0"
   },
   "dependencies": {
+    "buffer-to-stream": "1.0.0",
```
Comment (RayProud) [inline package.json:30]: Сделал как было в лекции (там использовали эту библиотеку) просто потому что очень много времени переписывал всё остальное. Просто не успеваю написать своё и протетсировать.

Можно оставить на конец.

File: src/commands/main.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ const writeFile = promisify(fs.writeFile);
 const colors = require(`colors/safe`);
 
 const {name} = require(`../../package.json`);
-const generateEntity = require(`../generate-entity`);
+const generateEntity = require(`../../test/mocks/generator/generate-entity`);
```
Comment (RayProud) [inline src/commands/main.js:8]: По аналогии с учебным проектом вынес в тесты/моки.

File: src/server/errors/index.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const NotImplementedError = require(`./not-implemented-error`);
+const BadRequestError = require(`./bad-request-error`);
+const NotFoundError = require(`./not-found-error`);
+
+module.exports = {
```
Comment (RayProud) [inline src/server/errors/index.js:5]: На работе привык делать в папках `index.js`, потому что потом удобно делать деструктуризацию из папки. Не надо к каждому отдельному файлу прописывать require.

File: src/server/index.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+const app = express();
+
+const {offersStore, imagesStore} = require(`./stores`);
+const offersRouter = require(`./routes/offers`)(offersStore, imagesStore);
+const defaultRouter = require(`./routes/default`);
+
+app.disable(`x-powered-by`);
+
+app.use(`/api/offers`, offersRouter);
+app.use(`/`, defaultRouter);
```
Comment (RayProud) [inline src/server/index.js:13]: Решил максимально разнести логику на роуты. 
Теперь этот файл действительно просто команда, которая запускает экспресс с несколькими роутами.

А вот сами роуты можно спокойно тестировать (что я ниже и сделал).

File: src/server/middlewares/error.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+const {generateHtmlError, generateJSONError} = require(`../errors-formatter`);
+
+module.exports = (router) => {
+  const ERROR_HANDLER = (err, req, res, _next) => {
+    if (err) {
+      const doesAcceptHtml = req.accepts(`html`) === `html`;
+      const errorMsg = doesAcceptHtml ? generateHtmlError(err) : generateJSONError(err);
+
+      if (err instanceof MongoError) {
+        res.status(400).json(err.message);
```
Comment (RayProud) [inline src/server/middlewares/error.js]: Честно говоря, не знаю, как мне это обрабатывать.

В идеале, нужно несколько ошибок обрабатывать по кодам ошибок Монги, чтобы была какая-то смысловая нагрузка.
Но на это пока просто не хватило времени. Плюс, у меня нет кейса, когда нельзя добавить два одинаковых объявления — это будут два разные объявления, хоть и с похожими данными.

Если есть предложения (и кейсы того, что может упасть) — с радостью сделаю. Но лучше попозже 🙏

File: src/server/routes/offers/default.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+const validate = require(`../../validation/validate`);
+const {postOfferScheme, getOffersScheme} = require(`../../validation/schemes`);
+const {asyncMiddleware, getRandomItemFromArray} = require(`../../../helpers/common`);
+
+const jsonParser = express.json();
+
+const DEFAULT_SKIP_AMOUNT = 0;
+const DEFAULT_LIMIT_AMOUNT = 20;
+const DEFAULT_NAMES = [`Keks`,
+  `Pavel`,
```
Comment (RayProud) [inline src/server/routes/offers/default.js:16]: Прочитал в ТЗ, что имя должно автоматически проставляться, если не пришло.
Вроде больше нигде не нужно, поэтому оставил этот список здесь.

File: src/server/routes/offers/default.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    skip,
+    total: await cursor.count(),
+  };
+};
+
+function prepareOfferForSave(offer) {
+  const date = Math.floor(Date.now() / 1000);
+
+  const offerToSave = Object.assign(
+      {},
```
Comment (RayProud) [inline src/server/routes/offers/default.js:39]: Только чистые функции! Никаких мутаций пришедших объектов, чтобы избежать неожиданностей!
Поэтому `Object.assign`.

О том, почему нужна эта функция — пишу дальше.

File: src/server/routes/offers/default.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    if (file) {
+      const {originalname, mimetype} = file;
+      body.avatar = {
+        name: originalname,
+        mimetype
+      };
+    }
+
+    const validatedOffer = validate(body, postOfferScheme);
+    const {address} = body;
```
Comment (RayProud) [inline src/server/routes/offers/default.js]: Кажется, что тут может упасть деструктуризация, но на самом деле `validate` быстрее выкинет ошибку, потому что это поле обязаьтельное. И если его нет, то до этого момента не дойдёт.

File: src/server/routes/offers/default.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    const [x, y] = address.split(`,`);
+    const location = {
+      x: +x.trim(),
+      y: +y.trim(),
+    };
+    validatedOffer.location = location;
```
Comment (RayProud) [inline src/server/routes/offers/default.js]: Согласно ТЗ, если валидация прошла, то в ответ мы должны вернуть тот же объект, что нам прислали, но с полем location.

File: src/server/routes/offers/default.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+    const objectToSave = prepareOfferForSave(validatedOffer);
```
Comment (RayProud) [inline src/server/routes/offers/default.js]: А вот в базе, согласно той же документации, мы уже должны хранить объекты с другой структурой. Где отдельно есть поля `date` и `author`, в добавок к тем, что есть. Поэтому такое разбиение.

Пример Request:
GET /offers/1519136255300:
```
{
  "author": {
    "name": "Pavel",
    "avatar": "api/offers/1519136255300/avatar"
  },
  "offer": {
    "title": "Маленькая квартирка рядом с парком",
    "address": "471, 545",
    "description": "Маленькая чистая квратира на краю парка. Без интернета, регистрации и СМС.",
    "price": 30000,
    "type": "flat",
    "rooms": 1,
    "guests": 1,
    "checkin": "9:00",
    "checkout": "7:00",
    "features": ["elevator", "conditioner"]
  },
  "location": {
    "x": 471,
    "y": 545
  },
  "date": 1519136255300
}
```

File: src/server/routes/offers/default.js (hunk 9 above)
Comment (RayProud) [reply src/server/routes/offers/default.js]: Пример:
Request:
POST /offers
```
{
  "name": "Pavel",
  "title": "Маленькая квартирка рядом с парком",
  "address": "570, 472",
  "description": "Маленькая чистая квратира на краю парка. Без интернета, регистрации и СМС.",
  "price": 30000,
  "type": "flat",
  "rooms": 1,
  "guests": 1,
  "checkin": "9:00",
  "checkout": "7:00",
  "features": ["elevator", "conditioner"]
}
```

Response:
200 OK
```
{
  "name": "Pavel",
  "title": "Маленькая квартирка рядом с парком",
  "address": "570, 472",
  "description": "Маленькая чистая квратира на краю парка. Без интернета, регистрации и СМС.",
  "price": 30000,
  "type": "flat",
  "rooms": 1,
  "guests": 1,
  "checkin": "9:00",
  "checkout": "7:00",
  "features": ["elevator", "conditioner"],
  "location": {
    "x": 570,
    "y": 472
  }
}
```

File: src/server/stores/images-store.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    return this._bucket;
+  }
+
+  async get(filename) {
+    const bucket = await this.getBucket();
+    const results = await bucket.find({filename}).toArray();
+    const entity = results[0];
+
+    if (!entity) {
+      return void 0;
```
Comment (RayProud) [inline src/server/stores/images-store.js:26]: Из лекции.

File: src/server/stores/images-store.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+class ImageStore {
+  async getBucket() {
+    if (this._bucket) {
+      return this._bucket;
+    }
+
+    const dBase = await db;
+
+    this._bucket = new mongodb.GridFSBucket(dBase, {
+      chunkSizeBytes: 512 * 1012,
```
Comment (RayProud) [inline src/server/stores/images-store.js:13]: Здесь магические числа тоже из лекции.
Можно в константы вынести, наверное.

File: src/server/validation/rules.js [hunk 13]
```diff
@@ -25,12 +25,6 @@ module.exports = {
     return null;
   },
   testType(value, type) {
-    // in case of multipart/form-data everything is a string
-    // so we should check if it's numeric and should be numeric
-    if (type === `number` && isNumeric(value)) {
```
Comment (RayProud) [inline src/server/validation/rules.js:30]: Вынес эту логику в `validate`. Кажется, здесь лучше просто проверять, а не преобразовывать и додумывать.

File: src/server/validation/schemes/post-offer.js [hunk 14]
```diff
@@ -1,67 +1,77 @@
...
-    testType: `string`,
     isRequired: true,
+    testType: `string`,
     testLength: {
       min: 30,
       max: 140,
     }
   },
-  type: {
+  description: {
```
Comment (RayProud) [inline src/server/validation/schemes/post-offer.js:10]: Почему-то в доке нет указаний про это поле.
Оно есть в примерах `GET`, но нет ничего про его ограничения. Додумал сам.

File: src/server/validation/schemes/post-offer.js [hunk 15]
```diff
@@ -1,67 +1,77 @@
...
-    testType: `number`,
     isRequired: true,
+    testType: `number`,
     isInRange: {
       min: 1,
       max: 100000,
     }
   },
+  guests: {
+    testType: `number`,
```
Comment (RayProud) [inline src/server/validation/schemes/post-offer.js:31]: Про `guests` почему-то тоже нет требований. Дописал как придумал.

File: src/server/validation/schemes/post-offer.js [hunk 16]
```diff
@@ -1,67 +1,77 @@
...
   address: {
-    testType: `string`,
     isRequired: true,
+    testType: `string`,
     testLength: {
       min: 0,
       max: 100,
     },
-    testMask: /^\d+,\s*\d+$/gi
+    testMask: /^\d+,\s*\d+$/
```
Comment (RayProud) [inline src/server/validation/schemes/post-offer.js:44]: Убрал флаги, потому что `i` для чисел роли не играет, а только путает. 
А вот `g` заставляет регулярку запоминат lastIndex и у меня ломается проверка.

File: test/default-route.test.js [hunk 17]
```diff
@@ -0,0 +1,45 @@
+const request = require(`supertest`);
+const assert = require(`assert`);
+const express = require(`express`);
+
+const defaultRouter = require(`../src/server/routes/default`);
+
+const app = express();
+app.use(`/`, defaultRouter);
```
Comment (RayProud) [inline test/default-route.test.js:8]: Сделал отдельные тесты на `/`. А те, что были, теперь тестируют `api/offers`.

File: src/server/validation/validate.js [hunk 18]
```diff
@@ -0,0 +1,64 @@
+const {BadRequestError} = require(`../errors`);
+const validateRules = require(`./rules`);
+const {isNumeric, isNil} = require(`../../helpers/common`);
+
+function prepareForValidation(value, currentScheme) {
+  // in case of multipart/form-data everything is a string
+  // so we should check if it's numeric and should be numeric
+  if (currentScheme.testType === `number` && isNumeric(value)) {
```
Comment (RayProud) [inline src/server/validation/validate.js:8]: В комментарии попытался описать потребность в этой функции.

File: src/server/validation/validate.js [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+    innerErrors = validationErrors;
+
+    return prev.concat(innerErrors);
+  }, []);
+
+  if (errors.length) {
+    throw new BadRequestError(errors);
+  }
+
+  return validatedObj;
```
Comment (RayProud) [inline src/server/validation/validate.js:61]: Теперь возвращаю отсюда объект (как и предлагали на лекциях).
Так удобнее записывать в базу уже отформатированные объекты, без лишних полей, которых нет в схеме, и в том типе, который должен быть, а не как из multipart/form-data приходит. Кстати, на этот баг наткнулся благодаря тестам! Это круто, я считаю.

File: test/api-offers.test.js [hunk 20]
```diff
@@ -1,7 +1,13 @@
 const request = require(`supertest`);
 const assert = require(`assert`);
+const express = require(`express`);
 
-const app = require(`../src/commands/server`).execute();
+const offersStoreMock = require(`./mocks/stores/offers-store-mock`);
+const imagesStoreMock = require(`./mocks/stores/images-store-mock`);
+const offersRouter = require(`../src/server/routes/offers`)(offersStoreMock, imagesStoreMock);
```
Comment (RayProud) [inline test/api-offers.test.js:7]: Теперь в тестах тестируются отдельные роуты.

File: test/api-offers.test.js [hunk 21]
```diff
@@ -261,99 +249,373 @@ describe(`GET /api/offers/:date`, () => {
...
-        ]
-      },
-      location: {
-        x: 310,
-        y: 450
-      },
-      date: 1538484899
-    };
+  describe(`with valid data`, () => {
+    describe(`of JSON`, () => {
```
Comment (RayProud) [inline test/api-offers.test.js:253]: Немного поменял схему describe, чтобы легче было дебажить.

File: src/server/validation/validate.js [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+  const validatedObj = {};
+
+  const errors = Object.keys(scheme).reduce((prev, key) => {
+    let innerErrors = [];
+    const currentScheme = scheme[key];
+
+    const value = prepareForValidation(objToValidate[key], currentScheme);
+
+    // no need to test a nil value which is not required
+    if (isNil(value) && !currentScheme.isRequired) {
```
Comment (RayProud) [inline src/server/validation/validate.js:30]: Оказывается, у меня всё это время не работала проверка на `isRequired`. Пофиксил.
В тестах тоже много переписал.

File: test/api-offers.test.js [hunk 23]
```diff
@@ -261,99 +249,373 @@ describe(`GET /api/offers/:date`, () => {
...
+      it(`with name field responds with a valid offer including a location field`, () => {
+        const mockRequestOffer = {
+          name: `Roman`,
```
Comment (RayProud) [inline test/api-offers.test.js:256]: Объекты все отличаются, поэтому в каждом `it` завожу новый.

Сперва хотел мутировать их внутри `it`, но вроде так делать нельзя — держать какую-то логику в тестах. Поэтому сделал максимально топорно.

File: test/api-offers.test.js [hunk 24]
```diff
@@ -898,40 +1503,82 @@ describe(`POST api/offers`, () => {
...
-describe(`POST api/offers/randomUrl`, () => {
-  const notImplementedUrl = `/api/offers/randomUrl`;
+describe(`GET api/offers/:data/avatar`, () => {
+  const realOfferDateWithAvatar = 1538317798;
+  const realOfferDateWithoutAvatar = 1538544650;
+  const notExistingOfferDate = 1511110629;
+
+  // how to test it with no real DB?
+  describe.skip(`with a valid date which has an avatar`, () => {
+    it(`responds with an image`, async () => {
```
Comment (RayProud) [inline test/api-offers.test.js:1523]: Вот что я не понял — как тестировать на моках добавление картинки? Поэтому заскипал.
Мне кажется, что с тестовой базой (вместо моков) эти тесты имели бы больший смысл.

File: src/server/middlewares/error.js (hunk 5 above)
Comment (TheDarkestDay) [reply src/server/middlewares/error.js]: Особо тут кейсы не придумаешь - у нас всего одна сущность, ни в каких отношениях она не состоит, и валидация всех полей осуществляется перед записью. Соответственно, у нас остается пласт ошибок в работе с самой Mongo:  проблемы с соединением, нехватка места/прав/чего-нибудь еще. Поэтому, лучше здесь использовать код 500, потому что природа этой ошибки чисто серверная - от клиента она никак не зависит.

File: src/database/db.js [hunk 25]
```diff
@@ -0,0 +1,13 @@
+const {MongoClient} = require(`mongodb`);
+
+const url = `mongodb://localhost:27017`;
```
Comment (TheDarkestDay) [inline src/database/db.js]: `url` тоже можно записать в константном регистре.

File: src/server/routes/offers/default.js [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+    const from = +skip || DEFAULT_SKIP_AMOUNT;
+    const till = +limit || DEFAULT_LIMIT_AMOUNT;
+
+    const page = await toPage(await offersRouter.offersStore.getAllOffers(), from, till);
+
+    res.json(page);
+  }));
+
+  offersRouter.post(`/`, jsonParser, upload.single(`avatar`), asyncMiddleware(async (req, res) => {
```
Comment (TheDarkestDay) [inline src/server/routes/offers/default.js]: Заметил, что не работает загрузка изображения в поле `preview`.

File: src/server/routes/offers/default.js (hunk 6 above)
Comment (TheDarkestDay) [reply src/server/routes/offers/default.js:16]: Хорошо.

**Review by TheDarkestDay — CHANGES_REQUESTED** [review]

File: test/api-offers.test.js (hunk 24 above)
Comment (TheDarkestDay) [reply test/api-offers.test.js:1523]: Нужно сконфигурировать заглушку так, чтобы она отдала нужное изображение, если её кто-нибудь попросит. Например, можно просто завести объект, который бы хранил в себе тестовые аватары для тестовых объявлений. Этот объект внутри себя использовал бы `ImageStoreMock` для того, чтобы эмулировать отдачу нужных картинок.

File: src/server/routes/offers/default.js (hunk 26 above)
Comment (RayProud) [reply src/server/routes/offers/default.js]: Добавил.

File: src/database/db.js (hunk 25 above)
Comment (RayProud) [reply src/database/db.js]: Поменял.

File: src/server/middlewares/error.js (hunk 5 above)
Comment (RayProud) [reply src/server/middlewares/error.js]: Поменял на 500.
